src/xmipp/libraries/data/monogenic.cpp

```diff
@@ -0,0 +1,712 @@
+	{
+		if (fourier_idx > 0){
+			FFT_IDX2DIGFREQ(fourier_idx - 1, volsize, freqL);
+		}
+		else{
+			freqL = sampling/(resolution + step);
+		}
+	}
+
+}
+
+
+// AMPLITUDEMONOSIG3D_LPF: Estimates the monogenic amplitude of a HPF map, myfftV is the
+// Fourier transform of that map. The monogenic amplitude is "amplitude", and the
+// parameters "fftVRiesz", "fftVRiesz_aux", "VRiesz", are auxiliar variables that will
+// be defined inside the function (they are input to speed up monores algorithm).
+// Freq, freqL and freqH, are the frequency of the HPF and the tails of a raise cosine.
+// Note that once the monogenic amplitude is estimated, it is low pass filtered to obtain
+// a smooth version and avoid rippling. freq_fourier_x, freq_fourier_y, freq_fourier_z,
+// are the accesible Fourier frequencies along each axis.
+void Monogenic::amplitudeMonoSig3D_LPF(const MultidimArray< std::complex<double> > &myfftV,
+		FourierTransformer &transformer_inv,
+		MultidimArray< std::complex<double> > &fftVRiesz,
+		MultidimArray< std::complex<double> > &fftVRiesz_aux, MultidimArray<double> &VRiesz,
+		double freq, double freqH, double freqL, MultidimArray<double> &iu,
+		Matrix1D<double> &freq_fourier_x, Matrix1D<double> &freq_fourier_y,
+		Matrix1D<double> &freq_fourier_z, MultidimArray<double> &amplitude,
+		int count, FileName fnDebug)
+{
+//	FourierTransformer transformer_inv;
+
+	fftVRiesz.initZeros(myfftV);
+	fftVRiesz_aux.initZeros(myfftV);
+	std::complex<double> J(0,1);
+
+	// Filter the input volume and add it to amplitude
+	long n=0;
+	double ideltal=PI/(freq-freqH);
+
+	for(size_t k=0; k<ZSIZE(myfftV); ++k)
+	{
+		for(size_t i=0; i<YSIZE(myfftV); ++i)
+		{
+			for(size_t j=0; j<XSIZE(myfftV); ++j)
+			{
+				double iun=DIRECT_MULTIDIM_ELEM(iu,n);
+				double un=1.0/iun;
+				if (freqH<=un && un<=freq)
+				{
+					DIRECT_MULTIDIM_ELEM(fftVRiesz, n) = DIRECT_MULTIDIM_ELEM(myfftV, n);
+					DIRECT_MULTIDIM_ELEM(fftVRiesz, n) *= 0.5*(1+cos((un-freq)*ideltal));//H;
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) = -J;
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) *= DIRECT_MULTIDIM_ELEM(fftVRiesz, n);
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) *= iun;
+				} else if (un>freq)
+				{
+					DIRECT_MULTIDIM_ELEM(fftVRiesz, n) = DIRECT_MULTIDIM_ELEM(myfftV, n);
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) = -J;
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) *= DIRECT_MULTIDIM_ELEM(fftVRiesz, n);
+					DIRECT_MULTIDIM_ELEM(fftVRiesz_aux, n) *= iun;
+				}
+				++n;
+			}
+		}
+	}
+
+	transformer_inv.inverseFourierTransform(fftVRiesz, amplitude);
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(amplitude)
+		DIRECT_MULTIDIM_ELEM(amplitude,n) *= DIRECT_MULTIDIM_ELEM(amplitude,n);
+
+	//TODO: create a macro with these kind of code
+	// Calculate first component of Riesz vector
+	double ux;
+	n=0;
+	for(size_t k=0; k<ZSIZE(myfftV); ++k)
+	{
+		for(size_t i=0; i<YSIZE(myfftV); ++i)
+		{
+			for(size_t j=0; j<XSIZE(myfftV); ++j)
+			{
+				ux = VEC_ELEM(freq_fourier_x,j);
```

Comment:
this can be extracted in template method

Reply:
I agree, but let's keep it for now